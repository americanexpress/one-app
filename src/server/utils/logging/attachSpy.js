/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export default function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function monkeypatchedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}
