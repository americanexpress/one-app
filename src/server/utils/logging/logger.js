/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import deepmerge from 'deepmerge';
import { argv } from 'yargs';
import { pino, multistream } from 'pino';
import productionConfig from './config/production';
import otelConfig, {
  createOtelTransport,
} from './config/otel';
import baseConfig from './config/base';

export function createLogger() {
  const useProductionConfig = !!(argv.logFormat === 'machine' || process.env.NODE_ENV !== 'development');

  let transport;

  if (process.env.NODE_ENV === 'development' && process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) {
    // Temporary solution until https://github.com/Vunovati/pino-opentelemetry-transport/issues/20 is resolved
    // eslint-disable-next-line global-require -- do not load development logger in production
    transport = multistream([{ stream: require('./config/development').default }, createOtelTransport()]);
  } else if (!useProductionConfig && !process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) {
    // eslint-disable-next-line global-require -- do not load development logger in production
    transport = require('./config/development').default;
  } else if (process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) {
    transport = createOtelTransport();
  }

  return pino(
    deepmerge(
      baseConfig,
      process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT
        ? otelConfig
        : useProductionConfig && productionConfig
    ),
    transport
  );
}

export default createLogger();
