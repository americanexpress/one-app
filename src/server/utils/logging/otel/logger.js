/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import os from 'node:os';
import {
  LoggerProvider,
  ConsoleLogRecordExporter,
  SimpleLogRecordProcessor,
  BatchLogRecordProcessor,
} from '@opentelemetry/sdk-logs';
import { OTLPLogExporter } from '@opentelemetry/exporter-logs-otlp-grpc';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { logs } from '@opentelemetry/api-logs';
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { ExpressInstrumentation } from '@opentelemetry/instrumentation-express';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import formatter from './formatter';
import readJsonFile from '../../readJsonFile';

const { buildVersion: version } = readJsonFile('../../../.build-meta.json');

const logMethods = ['error', 'warn', 'log', 'info', 'debug'];

let batchLogProcessor;

export function shutdownOtelLogger() {
  if (batchLogProcessor) return batchLogProcessor.shutdown();
  return undefined;
}

function setupTracer() {
  const provider = new NodeTracerProvider();
  provider.register();
  registerInstrumentations({
    instrumentations: [
      new HttpInstrumentation(),
      new ExpressInstrumentation(),
    ],
  });
}

export function createOtelLogger() {
  const customResourceAttributes = process.env.OTEL_RESOURCE_ATTRIBUTES ? process.env.OTEL_RESOURCE_ATTRIBUTES.split(';').reduce((acc, curr) => {
    const [key, value] = curr.split('=');
    return {
      ...acc,
      [key]: value,
    };
  }, {}) : {};

  const resource = new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: process.env.OTEL_SERVICE_NAME,
    [SemanticResourceAttributes.SERVICE_NAMESPACE]: process.env.OTEL_SERVICE_NAMESPACE,
    [SemanticResourceAttributes.SERVICE_INSTANCE_ID]: os.hostname(),
    [SemanticResourceAttributes.SERVICE_VERSION]: version,
    ...customResourceAttributes,
  });

  const collectorOptions = {
    headers: {},
    concurrencyLimit: 10,
    resource,
  };

  const loggerProvider = new LoggerProvider({ resource });
  const logExporter = new OTLPLogExporter(collectorOptions);

  if (process.env.NODE_ENV === 'development') {
    loggerProvider.addLogRecordProcessor(
      new SimpleLogRecordProcessor(new ConsoleLogRecordExporter())
    );
  }

  batchLogProcessor = new BatchLogRecordProcessor(logExporter);

  loggerProvider.addLogRecordProcessor(batchLogProcessor);
  logs.setGlobalLoggerProvider(loggerProvider);

  const otelLogger = logs.getLogger(process.env.OTEL_SERVICE_NAME);

  // Add trace IDs to logs
  setupTracer();

  const logger = logMethods.reduce((loggerObj, logMethod) => {
    function emitLog(...args) {
      const logRecord = formatter(logMethod, ...args);
      otelLogger.emit(logRecord);
    }
    return { ...loggerObj, [logMethod]: emitLog };
  }, {});

  return logger;
}

export function replaceGlobalConsoleWithOtelLogger(logger) {
  logMethods.forEach((methodName) => { console[methodName] = logger[methodName]; });
  return logger;
}
