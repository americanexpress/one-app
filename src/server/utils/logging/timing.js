/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// some timing helpers
function startTimer(obj) {
  // attach the data to the request for later use, similar to
  // https://github.com/expressjs/morgan/blob/1533cdcc512f4daf20a93cb0ab3b18d6563a897f/index.js#L491
  /* eslint-disable no-param-reassign, no-underscore-dangle */
  obj._startAt = process.hrtime();
  obj._startTime = new Date().toISOString();
  /* eslint-enable no-param-reassign, no-underscore-dangle */
}

function measureTime(prev) {
  // see `startTimer(...)` for justification
  const previousHrtime = prev._startAt; // eslint-disable-line no-underscore-dangle
  if (!Array.isArray(previousHrtime)) {
    return undefined;
  }
  const diff = process.hrtime(previousHrtime);
  return (diff[0] * 1e3) + (diff[1] * 1e-6); // ms
}

export {
  startTimer,
  measureTime,
};
