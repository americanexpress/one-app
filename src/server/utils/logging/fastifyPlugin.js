/*
 * Copyright 2022 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';
import fp from 'fastify-plugin';

import logger from './logger';

/*
TIMERS
`$RequestFullDuration` measures how long it took the whole request to be resolved
`$RequestOverhead` measures how long it took from when the request was made
                   until the route handler was called
`$RouteHandler` measures how long the route handler took to provide an output
`$ResponseBuilder` measures how long it took for fastify to build the response
                   after the payload/output from route handled was provided
Why symbols?
It prevents conflicts with existing keys, native or from other plugins
*/
export const $RequestFullDuration = Symbol('$RequestFullDuration');
export const $RequestOverhead = Symbol('$RequestOverhead');
export const $RouteHandler = Symbol('$RouteHandler');
export const $ResponseBuilder = Symbol('$ResponseBuilder');

const UTILS = {
  configureRequestLog: ({ log }) => log,
};

const getLocale = (req) => {
  // TODO: Verify if `store` is available
  if (req.store) {
    const state = req.store.getState();
    return state.getIn(['intl', 'activeLocale']);
  }

  return undefined;
};

const hrtimeToMs = (value) => (value[0] * 1e3) + (value[1] * 1e-6);

const startTimer = (obj, symbol) => {
  const time = process.hrtime();

  // eslint-disable-next-line no-param-reassign
  obj[symbol] = time;
};

const endTimer = (obj, symbol) => {
  const result = process.hrtime(obj[symbol]);
  const ms = hrtimeToMs(result);

  // eslint-disable-next-line no-param-reassign
  obj[symbol] = ms;

  return ms;
};

const buildMetaData = (request, reply) => {
  const { headers } = request;

  return {
    method: request.method,
    correlationId: headers['correlation-id'],
    // null to explicitly signal no value, undefined if not expected for every request
    host: headers.host || null,
    referrer: headers.referer || headers.referrer || null,
    userAgent: headers['user-agent'] || null,
    location: reply.getHeader('location') || undefined,
    forwarded: headers.forwarded || null,
    forwardedFor: headers['x-forwarded-for'] || null,
    locale: getLocale(request),
  };
};

const logClientRequest = (request, reply) => {
  const getTimer = (symbol) => request[symbol];
  const ttfb = getTimer($RequestFullDuration) - getTimer($ResponseBuilder);
  const log = {
    type: 'request',
    request: {
      direction: 'in',
      protocol: request.protocol,
      address: {
        uri: url.format({
          protocol: request.protocol,
          port: request.port,
          hostname: request.hostname,
          pathname: request.originalUrl,
        }),
      },
      metaData: buildMetaData(request, reply),
      timings: { // https://www.w3.org/TR/navigation-timing/
        // navigationStart: 0,
        // redirectStart: 0,
        // redirectEnd: 0,
        // fetchStart: 0,
        // domainLookupStart: 10,
        // domainLookupEnd: 20,
        // connectStart: 30,
        // secureConnectionStart: 40,
        // connectEnd: 50,
        // requestStart: 60,
        // requestEnd: 70, // optional? not part of the W3C spec
        // responseStart: 80,
        // responseEnd: 90,
        // FIXME: mimic the w3 timing names
        duration: Math.round(getTimer($RequestFullDuration)),
        ttfb: ttfb >= 0 ? Math.round(ttfb) : null,
        requestOverhead: Math.round(getTimer($RequestOverhead)),
        routeHandler: Math.round(getTimer($RouteHandler)),
        responseBuilder: Math.round(getTimer($ResponseBuilder)),
      },
      statusCode: reply.statusCode,
      statusText: reply.raw.statusMessage || undefined,
    },
  };

  const configuredLog = UTILS.configureRequestLog({
    req: request.raw,
    res: reply.raw,
    log,
  });

  logger.info(configuredLog);
};

export const setConfigureRequestLog = (newConfigureRequestLog) => {
  UTILS.configureRequestLog = newConfigureRequestLog;
};

const fastifyPlugin = (fastify, _opts, done) => {
  // decorators should be initialized per Fastify documentation
  // see https://www.fastify.io/docs/latest/Reference/Decorators/#decoratereplyname-value-dependencies
  fastify.decorateRequest($RequestOverhead, null);
  fastify.decorateRequest($RequestFullDuration, null);
  fastify.decorateRequest($RouteHandler, null);
  fastify.decorateRequest($ResponseBuilder, null);

  // NOTE: this is needed for backward compatibility since
  //       we were exposing the 'req' and 'res' from ExpressJS
  //       to the App Config.
  fastify.addHook('onRequest', async (request, reply) => {
    request.raw.originalUrl = request.raw.url;
    request.raw.id = request.id;
    request.raw.hostname = request.hostname;
    request.raw.ip = request.ip;
    request.raw.ips = request.ips;
    request.raw.log = request.log;
    // eslint-disable-next-line no-param-reassign
    reply.raw.log = request.log;

    // backward compatibility for body-parser
    if (request.body) {
      request.raw.body = request.body;
    }
    // backward compatibility for cookie-parser
    if (request.cookies) {
      request.raw.cookies = request.cookies;
    }

    // Make it lazy as it does a bit of work
    Object.defineProperty(request.raw, 'protocol', {
      get() {
        return request.protocol;
      },
    });
  });

  fastify.addHook('onRequest', async (request) => {
    startTimer(request, $RequestOverhead);
    startTimer(request, $RequestFullDuration);
  });

  fastify.addHook('preHandler', async (request) => {
    endTimer(request, $RequestOverhead);
    startTimer(request, $RouteHandler);
  });

  fastify.addHook('onSend', async (request, _reply, payload) => {
    endTimer(request, $RouteHandler);
    startTimer(request, $ResponseBuilder);

    // Note: `onSend` is meant to be used to modify the payload for the response.
    //        we use it to calculate TTFB (time to first byte) since it's called
    //        by fastify right before `writeHead` is called.
    return payload;
  });

  fastify.addHook('onResponse', async (request, reply) => {
    endTimer(request, $ResponseBuilder);
    // same as 'reply.getResponseTime()' but our approach
    // helps us to make the code cleaner
    endTimer(request, $RequestFullDuration);

    logClientRequest(request, reply);
  });

  done();
};

export default fp(fastifyPlugin, {
  fastify: '4.x',
  name: 'logging',
});
