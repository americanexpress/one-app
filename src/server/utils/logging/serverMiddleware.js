/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';

import { monkeypatches } from '@americanexpress/lumberjack';

import { startTimer, measureTime } from './timing';
import logger from './logger';

const passThrough = ({ log }) => log;
const tenantUtils = {
  configureRequestLog: passThrough,
};

function getLocale(req) {
  if (req.store) {
    const state = req.store.getState();
    return state.getIn(['intl', 'activeLocale']);
  }
  return undefined;
}

function buildMetaData(req, res) {
  const { headers } = req;
  return {
    method: req.method,
    correlationId: headers['correlation-id'],
    // null to explicitly signal no value, undefined if not expected for every request
    host: headers.host || null,
    referrer: headers.referer || headers.referrer || null,
    userAgent: headers['user-agent'] || null,
    location: res.getHeader('location') || undefined,
    forwarded: headers.forwarded || null,
    forwardedFor: headers['x-forwarded-for'] || null,
    locale: getLocale(req),
  };
}

function logClientRequest(req, res) {
  const fullDuration = measureTime(req);
  const responseDuration = measureTime(res);
  const ttfb = fullDuration - responseDuration;
  const log = {
    type: 'request',
    request: {
      direction: 'in',
      protocol: req.protocol,
      address: {
        uri: url.format({
          protocol: req.protocol,
          port: req.port,
          hostname: req.hostname,
          pathname: req.originalUrl,
        }),
      },
      metaData: buildMetaData(req, res),
      timings: { // https://www.w3.org/TR/navigation-timing/
        // navigationStart: 0,
        // redirectStart: 0,
        // redirectEnd: 0,
        // fetchStart: 0,
        // domainLookupStart: 10,
        // domainLookupEnd: 20,
        // connectStart: 30,
        // secureConnectionStart: 40,
        // connectEnd: 50,
        // requestStart: 60,
        // requestEnd: 70, // optional? not part of the W3C spec
        // responseStart: 80,
        // responseEnd: 90,
        // FIXME: mimic the w3 timing names
        duration: Math.round(fullDuration),
        ttfb: ttfb >= 0 ? Math.round(ttfb) : null,
      },
      statusCode: res.statusCode,
      statusText: res.statusMessage || undefined,
    },
  };

  const configuredLog = tenantUtils.configureRequestLog({ req, res, log });
  logger.info(configuredLog);
}

export const setConfigureRequestLog = (newConfigureRequestLog = passThrough) => {
  tenantUtils.configureRequestLog = newConfigureRequestLog;
};

export default function serverMiddleware(req, res, next) {
  startTimer(req);
  monkeypatches.attachSpy(res, 'writeHead', () => startTimer(res));
  res.on('close', () => logClientRequest(req, res));
  setImmediate(next);
}
