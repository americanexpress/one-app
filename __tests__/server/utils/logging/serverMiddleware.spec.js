/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { fromJS } from 'immutable';

import serverMiddleware, { setConfigureRequestLog } from '../../../../src/server/utils/logging/serverMiddleware';

import { startTimer, measureTime } from '../../../../src/server/utils/logging/timing';
import logger from '../../../../src/server/utils/logging/logger';

jest.mock('../../../../src/server/utils/logging/timing', () => ({
  startTimer: jest.fn(),
  measureTime: jest.fn(() => 12),
}));
jest.mock('../../../../src/server/utils/logging/logger', () => ({ info: jest.fn() }));

describe('serverMiddleware', () => {
  jest.spyOn(global, 'setImmediate').mockImplementation((cb) => cb());

  function createReqResNext() {
    const req = {
      method: 'GET',
      headers: {
        'correlation-id': '123',
        host: 'example.com',
        referer: 'https://example.com/other-place',
        'user-agent': 'Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)',
      },
      cookies: {},
      protocol: 'https',
      hostname: 'example.com',
      originalUrl: '/resource',
      res: {
        statusCode: 200,
        statusMessage: 'OK',
      },
    };
    const res = {
      privateHeaderStore: {},
      getHeader: jest.fn((name) => res.privateHeaderStore[name]),
      writeHead: jest.fn(),
      on: jest.fn(),
    };
    const next = jest.fn();

    return { req, res, next };
  }

  function findMockCallForEvent(objectWithOnMethod, eventName) {
    return objectWithOnMethod.on.mock.calls.reduce((a, c) => {
      if (c[0] === eventName) {
        return c[1];
      }
      return a;
    }, null);
  }

  it('starts the request timer when the request comes in', () => {
    startTimer.mockClear();
    const { req, res, next } = createReqResNext();
    serverMiddleware(req, res, next);
    expect(startTimer).toHaveBeenCalledTimes(1);
    expect(startTimer).toHaveBeenCalledWith(req);
  });

  it('calls the next middleware', () => {
    const { req, res, next } = createReqResNext();
    setImmediate.mockClear();
    serverMiddleware(req, res, next);
    expect(setImmediate).toHaveBeenCalledTimes(1);
    expect(setImmediate).toHaveBeenCalledWith(next);
    expect(next).toHaveBeenCalledTimes(1);
  });

  it('starts the response timer when headers are beginning to be written', () => {
    const { req, res, next } = createReqResNext();
    serverMiddleware(req, res, next);
    startTimer.mockClear();
    res.writeHead();
    expect(startTimer).toHaveBeenCalledTimes(1);
    expect(startTimer).toHaveBeenCalledWith(res);
  });

  it('logs the request & response when the response closes', () => {
    const { req, res, next } = createReqResNext();
    logger.info.mockClear();
    serverMiddleware(req, res, next);
    findMockCallForEvent(res, 'close')();
    expect(logger.info).toHaveBeenCalledTimes(1);
    expect(logger.info.mock.calls[0]).toMatchSnapshot();
  });

  describe('log entry', () => {
    it('uses the time to first byte', () => {
      const { req, res, next } = createReqResNext();
      logger.info.mockClear();
      measureTime
        .mockImplementationOnce(() => 56)
        .mockImplementationOnce(() => 34);
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.timings).toHaveProperty('ttfb', 22);
    });

    it('uses null for the time to first byte when not present', () => {
      const { req, res, next } = createReqResNext();
      logger.info.mockClear();
      measureTime
        .mockImplementationOnce(() => undefined)
        .mockImplementationOnce(() => undefined);
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.timings).toHaveProperty('ttfb', null);
    });

    it('uses the host header', () => {
      const { req, res, next } = createReqResNext();
      const expected = 'example.com';
      req.headers.host = expected;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('host', expected);
    });

    it('uses null for the host header when not present', () => {
      const { req, res, next } = createReqResNext();
      delete req.headers.host;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('host', null);
    });

    it('uses the referrer header', () => {
      const { req, res, next } = createReqResNext();
      const expected = 'https://example.com/referrer';
      req.headers.referrer = expected;
      delete req.headers.referer;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('referrer', expected);
    });

    it('uses the referer header for referrer', () => {
      const { req, res, next } = createReqResNext();
      const expected = 'https://example.com/referer';
      delete req.headers.referrer;
      req.headers.referer = expected;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('referrer', expected);
    });

    it('uses null for the referrer header when not present', () => {
      const { req, res, next } = createReqResNext();
      delete req.headers.referrer;
      delete req.headers.referer;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('referrer', null);
    });

    it('uses the user agent header', () => {
      const { req, res, next } = createReqResNext();
      const expected = 'Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)';
      req.headers['user-agent'] = expected;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('userAgent', expected);
    });

    it('uses null for the user agent header when not present', () => {
      const { req, res, next } = createReqResNext();
      delete req.headers['user-agent'];
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('userAgent', null);
    });

    it('uses the location header', () => {
      const { req, res, next } = createReqResNext();
      const expected = 'https://example.com/another-place';
      res.privateHeaderStore.location = expected;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('location', expected);
    });

    it('uses undefined for the location header when not present', () => {
      const { req, res, next } = createReqResNext();
      delete res.privateHeaderStore.location;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('location', undefined);
    });

    it('uses the forwarded header', () => {
      const { req, res, next } = createReqResNext();
      const expected = 'by=example;for=example;host=example.com;proto=https';
      req.headers.forwarded = expected;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('forwarded', expected);
    });

    it('uses null for the forwarded header when not present', () => {
      const { req, res, next } = createReqResNext();
      delete req.headers.forwarded;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('forwarded', null);
    });

    it('uses the x-forwarded-for header', () => {
      const { req, res, next } = createReqResNext();
      const expected = '10.20.30.40';
      req.headers['x-forwarded-for'] = expected;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('forwardedFor', expected);
    });

    it('uses undefined for the x-forwarded-for header when not present', () => {
      const { req, res, next } = createReqResNext();
      delete req.headers['x-forwarded-for'];
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('forwardedFor', null);
    });

    it('uses the activeLocale when the redux store is on the request', () => {
      const { req, res, next } = createReqResNext();
      const expected = 'tlh';
      req.store = { getState: () => fromJS({ intl: fromJS({ activeLocale: expected }) }) };
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('locale', expected);
    });

    it('uses undefined as the activeLocale when the redux store is not on the request', () => {
      const { req, res, next } = createReqResNext();
      delete req.store;
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('locale', undefined);
    });
  });

  describe('setConfigureRequestLog', () => {
    const customConfigureRequestLog = jest.fn(() => ({ log }) => log);

    beforeEach(() => {
      customConfigureRequestLog.mockClear();
    });

    it('updates configureRequestLog function', () => {
      setConfigureRequestLog(customConfigureRequestLog);
      const { req, res, next } = createReqResNext();
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      expect(customConfigureRequestLog.mock.calls[0][0].req).toEqual(req);
      expect(customConfigureRequestLog.mock.calls[0][0].res).toEqual(res);
      expect(customConfigureRequestLog.mock.calls[0][0].log).toMatchSnapshot();
    });

    it('sets to default passthrough when no arg given', () => {
      setConfigureRequestLog(customConfigureRequestLog);
      setConfigureRequestLog();
      const { req, res, next } = createReqResNext();
      logger.info.mockClear();
      serverMiddleware(req, res, next);
      findMockCallForEvent(res, 'close')();
      expect(logger.info).toHaveBeenCalledTimes(1);
      expect(customConfigureRequestLog).not.toHaveBeenCalled();
    });
  });
});
