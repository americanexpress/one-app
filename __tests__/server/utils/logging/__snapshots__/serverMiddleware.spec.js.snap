// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serverMiddleware logs the request & response when the response closes 1`] = `
Array [
  Object {
    "request": Object {
      "address": Object {
        "uri": "https://example.com/resource",
      },
      "direction": "in",
      "metaData": Object {
        "correlationId": "123",
        "forwarded": null,
        "forwardedFor": null,
        "host": "example.com",
        "locale": undefined,
        "location": undefined,
        "method": "GET",
        "referrer": "https://example.com/other-place",
        "userAgent": "Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)",
      },
      "protocol": "https",
      "statusCode": undefined,
      "statusText": undefined,
      "timings": Object {
        "duration": 12,
        "ttfb": 0,
      },
    },
    "type": "request",
  },
]
`;

exports[`serverMiddleware setConfigureRequestLog updates configureRequestLog function 1`] = `
Object {
  "request": Object {
    "address": Object {
      "uri": "https://example.com/resource",
    },
    "direction": "in",
    "metaData": Object {
      "correlationId": "123",
      "forwarded": null,
      "forwardedFor": null,
      "host": "example.com",
      "locale": undefined,
      "location": undefined,
      "method": "GET",
      "referrer": "https://example.com/other-place",
      "userAgent": "Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)",
    },
    "protocol": "https",
    "statusCode": undefined,
    "statusText": undefined,
    "timings": Object {
      "duration": 12,
      "ttfb": 0,
    },
  },
  "type": "request",
}
`;
